File: solutions/julia/triangle/1/triangle.jl
# Helper function to check if the side lengths in the collection form a valid triangle
"""
    is_valid_triangle(sides::Vector{<:Real})::Bool

Checks if the given `sides` (a vector of lengths) can form a valid triangle.
Returns `true` if valid, `false` otherwise.
"""
function is_valid_triangle(sides::Vector{<:Real})::Bool
    # A triangle must have exactly three sides.
    if length(sides) != 3
        return false
    end

    # Assign sides to variables for clarity.
    a, b, c = sides[1], sides[2], sides[3]

    # 1. All sides must be of length > 0
    if a <= 0 || b <= 0 || c <= 0
        return false
    end

    # 2. Triangle Inequality Theorem: The sum of any two sides must be
    #    greater than or equal to the third side.
    if !(a + b >= c && b + c >= a && a + c >= b)
        return false
    end

    return true
end

"""
    is_equilateral(sides::Vector{<:Real})::Bool

Determines if a triangle with given `sides` is equilateral.
Returns `true` if it's a valid triangle and all three sides are equal.
"""
function is_equilateral(sides::Vector{<:Real})::Bool
    # First, ensure it's a valid triangle.
    if !is_valid_triangle(sides)
        return false
    end

    # Assign sides to variables for clarity.
    a, b, c = sides[1], sides[2], sides[3]
    
    # An equilateral triangle has all three sides equal.
    return a == b && b == c
end

"""
    is_isosceles(sides::Vector{<:Real})::Bool

Determines if a triangle with given `sides` is isosceles.
Returns `true` if it's a valid triangle and at least two sides are equal.
Note: An equilateral triangle is also considered isosceles by this definition.
"""
function is_isosceles(sides::Vector{<:Real})::Bool
    # First, ensure it's a valid triangle.
    if !is_valid_triangle(sides)
        return false
    end

    # Assign sides to variables for clarity.
    a, b, c = sides[1], sides[2], sides[3]
    
    # An isosceles triangle has at least two sides equal.
    return a == b || b == c || a == c
end

"""
    is_scalene(sides::Vector{<:Real})::Bool

Determines if a triangle with given `sides` is scalene.
Returns `true` if it's a valid triangle and all three sides are different lengths.
"""
function is_scalene(sides::Vector{<:Real})::Bool
    # First, ensure it's a valid triangle.
    if !is_valid_triangle(sides)
        return false
    end

    # Assign sides to variables for clarity.
    a, b, c = sides[1], sides[2], sides[3]
    
    # A scalene triangle has all sides of different lengths.
    return a != b && b != c && a != c
end

File: solutions/go/triangle/1/triangle.go
package triangle

// Kind represents the type of a triangle.
type Kind int

// These constants define the possible classifications for a triangle.
const (
	NaT Kind = iota // Not a triangle
	Equ             // Equilateral
	Iso             // Isosceles
	Sca             // Scalene
)

func KindFromSides(a, b, c float64) Kind {
	// Rule 1: All sides must be of length > 0.
	// If any side is non-positive, it's not a valid triangle.
	if a <= 0 || b <= 0 || c <= 0 {
		return NaT
	}

	// Rule 2: Triangle Inequality Theorem
	// The sum of the lengths of any two sides must be greater than or equal to the length of the third side.
	// This ensures the sides can actually form a closed shape.
	if (a+b < c) || (b+c < a) || (a+c < b) {
		return NaT
	}

	// Now that we know it's a valid triangle, apply classification rules.
	// We check from the most specific to the least specific.

	// Equilateral: All three sides are the same length.
	if a == b && b == c {
		return Equ
	}

	// Isosceles: At least two sides are the same length.
	// This check comes after Equilateral, so if it's Equ, it won't hit this.
	if a == b || b == c || a == c {
		return Iso
	}

	// Scalene: All sides are of different lengths.
	// If it passed the above checks (meaning it's not Equ or Iso), it must be Scalene.
	return Sca
}